meshmode/dof_array.py

```diff
@@ -294,23 +294,44 @@ class _EntryNotProvided:
 
 
 @serialize_container.register(DOFArray)
-def _(ary: DOFArray):
+def _serialize_container_dofarray(ary: DOFArray):
     return enumerate(ary._data)
 
 
-@deserialize_container_class.register(DOFArray)
-def _(cls, template: Any, iterable: Iterable[Tuple[Any, Any]], *,
-        actx: Optional[ArrayContext] = None):
-    iterable = list(iterable)
-    result = [_EntryNotProvided] * len(iterable)
+def _raise_index_inconsistency(i, stream_i):
+    raise ValueError("out-of-sequence indices supplied in DOFArray deserialization "
+            f"(expected {i}, received {stream_i})")
 
-    for i, subary in iterable:
-        result[i] = subary
 
-    if any(subary is _EntryNotProvided for subary in result):
-        raise ValueError("'iterable' does not contain all indices")
+@deserialize_container.register(DOFArray)
+def _deserialize_container_dofarray(
+        template: Any, iterable: Iterable[Tuple[Any, Any]]):
+    return type(template)(
+            template.array_context,
+            data=tuple(
+                v if i == stream_i else _raise_index_inconsistency(i, stream_i)
+                for i, (stream_i, v) in enumerate(iterable)))
 
-    return cls(actx, data=tuple(result))
+
+@_freeze.register(DOFArray)
+def _freeze_dofarray(ary, actx=None):
+    if actx is not None:
+        if actx is not ary.array_context:
+            raise ValueError("supplied array context does not agree with the one "
+                    "in the DOFArray in freeze(DOFArray)")
+    return type(ary)(
+        None,
+        tuple(_freeze(subary, ary.array_context) for subary in ary._data))
```

Comment:
Doe sthis need to call recursively call `_freeze`? `DOFArray` components should just be freezable by `actx.freeze`.

Reply:
f171e0b